keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,999 @@
+nominate the Pod based on the calculation,
+and actually makes the API calls in the goroutine.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### When kube-apiserver is unstable
+
+When kube-apiserver is unstable and API calls at the preemption goroutine fails frequently,
+the scheduler could make a non-best scheduling result 
+because the scheduler nominates pods at `PostFilter` though, those Pods won't be scheduled on nodes because the preemption API calls fail.
+
+Let's say many mid-priority Pods are making the preemption API calls.
+During the preemption goroutine for them are runnning, the scheduler assumes they'll be scheduled at the Nodes eventually
+that the preemptions are targeting via `.Status.NominatedNodeName`.
+So, other mid-priority or lower priority Pods' scheduling take those preempter Pods into consideration, 
+which is correct if the preemption goroutine finishes successful actually, while which results in non-best scheduling results.
+(Higher priority Pods won't be affected; Pods can take place of reserved for lower priority Pods via `.Status.NominatedNodeName`)
+
+But, in the first place though, when kube-apiserver is unstable, the scheduler doesn't behave well 
+because it works with a lot of communication with kube-apiserver.
+Even if the scheduler makes the best scheduling result, the binding API might fail after all.
+
+So, we don't have to pay a special attention to this issue.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+To achieve an asynchronous preemption, we will change the preemption plugin's implementation like the following:
+1. The preemption PostFilter plugin calculates the preemption target and nominate the Pod for the Node. (We'll use `AddNominatedPod` API exposed from the scheduling framework to plugins.)
+2. The preemption PostFilter plugin starts the goroutine to make API calls inside, and return success status (= not wait for the goroutine to finish).
+3. The preemption plugin gates the Pod, which the preemption is in-progress, at PreEnqueue extension point so that the target Pod won't be retried during the preemption.
```

Comment:
I don't think we should do this. The pod should be able to compete for spaced freed by other pods that it didn't preempt.

Reply:
Let's discuss this topic in this thread -
https://github.com/kubernetes/enhancements/pull/4833#discussion_r1782440928

Reply:
Because others might not see the other thread. I'm not in favor of this KEP with this restriction.

Reply:
```suggestion
3. The preemption plugin blocks the Pod while the preemption routine is in-progress, using PreEnqueue extension point, so that the target Pod won't be retried during this time.
```

And add a sentence to indicate why we do this.